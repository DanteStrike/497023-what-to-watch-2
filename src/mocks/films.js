export const films = [
  {
    id: 1,
    name: `filmOne`,
    posterImage: `img/filmOne-poster.jpg`,
    preview: {
      image: `img/filmOne.jpg`,
      videoSrc: `https://some-linkOne`
    },
    background: {
      image: `img/filmOne.jpg`,
      color: `#ffffff`
    },
    videoSrc: `https://some-linkOne`,
    description: `Some text filmOne`,
    rating: 1.0,
    scoresCount: 1,
    director: `directorOne`,
    starring: [`actorOne`, `actorTwo`],
    runTime: 1,
    genre: `genreOne`,
    released: 2019,
    isFavorite: false,
  }, {
    id: 3,
    name: `filmTwo`,
    posterImage: `img/filmTwo-poster.jpg`,
    preview: {
      image: `img/filmTwo.jpg`,
      videoSrc: `https://some-linkTwo`
    },
    background: {
      image: `img/filmTwo.jpg`,
      color: `#000000`
    },
    videoSrc: `https://some-linkTwo`,
    description: `Some text filmTwo`,
    rating: 2.0,
    scoresCount: 2,
    director: `directorTwo`,
    starring: [`actorThree`, `actorFour`],
    runTime: 2,
    genre: `genreTwo`,
    released: 2020,
    isFavorite: true,
  }
];
