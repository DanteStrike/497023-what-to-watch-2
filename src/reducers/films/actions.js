import types from "./types.js";
import utils from "./utils.js";


const loadFilms = (filmsRAW) => {
  return {
    type: types.LOAD_FILMS,
    payload: utils.transformFilmsRAW(filmsRAW)
  };
};

const loadPromo = (PromoRAW) => {
  return {
    type: types.LOAD_PROMO,
    payload: PromoRAW[`id`]
  };
};


export default {
  loadFilms,
  loadPromo
};
