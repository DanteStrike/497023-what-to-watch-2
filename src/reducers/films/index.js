export {default as filmsTypes} from "./types.js";
export {default as filmsActions} from "./actions.js";
export {default as filmsSelectors} from "./selectors.js";
export {default as filmsReducer} from "./reducers.js";
