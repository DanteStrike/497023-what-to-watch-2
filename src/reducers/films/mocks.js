const filmsRAW = [
  {
    "id": 1,
    "name": `filmOne`,
    "poster_image": `img/filmOne-poster.jpg`,
    "preview_image": `img/filmOne.jpg`,
    "background_image": `img/filmOne.jpg`,
    "background_color": `#ffffff`,
    "video_link": `https://some-linkOne`,
    "preview_video_link": `https://some-linkOne`,
    "description": `Some text filmOne`,
    "rating": 1.0,
    "scores_count": 1,
    "director": `directorOne`,
    "starring": [`actorOne`, `actorTwo`],
    "run_time": 1,
    "genre": `genreOne`,
    "released": 2019,
    "is_favorite": false,
  },
  {
    "id": 3,
    "name": `filmTwo`,
    "poster_image": `img/filmTwo-poster.jpg`,
    "preview_image": `img/filmTwo.jpg`,
    "background_image": `img/filmTwo.jpg`,
    "background_color": `#000000`,
    "video_link": `https://some-linkTwo`,
    "preview_video_link": `https://some-linkTwo`,
    "description": `Some text filmTwo`,
    "rating": 2.0,
    "scores_count": 2,
    "director": `directorTwo`,
    "starring": [`actorThree`, `actorFour`],
    "run_time": 2,
    "genre": `genreTwo`,
    "released": 2020,
    "is_favorite": true,
  }
];

const adaptedFilmsData = [
  {
    id: 1,
    name: `filmOne`,
    posterImage: `img/filmOne-poster.jpg`,
    preview: {
      image: `img/filmOne.jpg`,
      videoSrc: `https://some-linkOne`
    },
    background: {
      image: `img/filmOne.jpg`,
      color: `#ffffff`
    },
    videoSrc: `https://some-linkOne`,
    description: `Some text filmOne`,
    rating: 1.0,
    scoresCount: 1,
    director: `directorOne`,
    starring: [`actorOne`, `actorTwo`],
    runTime: 1,
    genre: `genreOne`,
    released: 2019,
    isFavorite: false,
  }, {
    id: 3,
    name: `filmTwo`,
    posterImage: `img/filmTwo-poster.jpg`,
    preview: {
      image: `img/filmTwo.jpg`,
      videoSrc: `https://some-linkTwo`
    },
    background: {
      image: `img/filmTwo.jpg`,
      color: `#000000`
    },
    videoSrc: `https://some-linkTwo`,
    description: `Some text filmTwo`,
    rating: 2.0,
    scoresCount: 2,
    director: `directorTwo`,
    starring: [`actorThree`, `actorFour`],
    runTime: 2,
    genre: `genreTwo`,
    released: 2020,
    isFavorite: true,
  }
];

const normalizedFilmsData = {
  byIDs: {
    "1": adaptedFilmsData[0],
    "3": adaptedFilmsData[1]
  },
  allIDs: [1, 3]
};

const store = {
  films: {
    data: normalizedFilmsData
  }
};

export default {
  filmsRAW,
  adaptedFilmsData,
  normalizedFilmsData,
  store
};
