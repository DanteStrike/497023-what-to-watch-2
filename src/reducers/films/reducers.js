import types from "./types.js";
import {combineReducers} from "redux";

const filmsReducer = (state = {}, action) => {
  if (action.type === types.LOAD_FILMS) {
    return action.payload;
  }

  return state;
};

const reducer = combineReducers({
  data: filmsReducer
});

export default reducer;
