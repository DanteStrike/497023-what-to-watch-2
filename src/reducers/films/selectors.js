import {createSelector} from "reselect";

const getAllIDs = (store) => store.films.data.allIDs;
const getFilmsByID = (store) => store.films.data.byID;

const getCardsInfo = createSelector(
    [getAllIDs, getFilmsByID],
    (allIDs, films) => allIDs.map((filmID) => ({
      id: films[filmID].id,
      name: films[filmID].name,
      preview: films[filmID].preview,
    }))
);


export default {
  getAllIDs,
  getFilmsByID,
  getCardsInfo
};
