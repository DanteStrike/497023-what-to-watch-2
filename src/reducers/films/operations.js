import {filmsActions} from "./films";

const loadFilms = () => (dispatch, _, api) => {
  return api.get(`/films`)
    .then((response) => {
      dispatch(filmsActions.loadFilms(response.data));
    });
};

const loadPromo = () => (dispatch, _, api) => {
  return api.get(`/films/promo`)
    .then((response) => {
      dispatch(filmsActions.loadPromo(response.data));
    });
};

export default {
  loadFilms,
  loadPromo
};
