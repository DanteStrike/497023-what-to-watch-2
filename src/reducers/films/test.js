import types from "./types.js";
import mocks from "./mocks.js";
import utils from "./utils.js";
import actions from "./actions.js";
import selectors from "./selectors.js";
import reducer from "./reducers.js";

describe(`Reducers: Films utils`, () => {
  describe(`Transform RAW server data`, () => {
    it(`Util adaptFilmRawData`, () => {
      const filmRAW = mocks.filmsRAW[0];
      const adaptedFilmData = mocks.adaptedFilmsData[0];

      expect(utils.adaptFilmRAW(filmRAW)).toEqual(adaptedFilmData);
    });

    it(`Util adaptFilmsRAW`, () => {
      expect(utils.adaptFilmsRAW(mocks.filmsRAW)).toEqual(mocks.adaptedFilmsData);
      expect(utils.adaptFilmsRAW([])).toEqual([]);
    });

    it(`Util normalizeFilms`, () => {
      expect(utils.normalizeFilms(mocks.adaptedFilmsData)).toEqual(mocks.normalizedFilmsData);
      expect(utils.normalizeFilms([])).toEqual({
        byIDs: {},
        allIDs: []
      });
    });

    it(`Util transformFilmsRAW`, () => {
      expect(utils.transformFilmsRAW(mocks.filmsRAW)).toEqual(mocks.normalizedFilmsData);
      expect(utils.transformFilmsRAW([])).toEqual({
        byIDs: {},
        allIDs: []
      });
    });
  });
});

describe(`Reducers: Films actions`, () => {
  it(`Action loadFilm`, () => {
    const spy = jest.spyOn(utils, `transformFilmsRAW`);
    spy.mockReturnValue([{film: `norm`}]);
    expect(actions.loadFilms([{film: `any`}])).toEqual({
      type: types.LOAD_FILMS,
      payload: [{film: `norm`}]
    });
  });
});

describe(`Reducers: Films reducers`, () => {
  describe(`Reducer filmsReducer`, () => {
    const initState = {
      data: {
        some: `any`
      }
    };

    it(`Should set state on correct action`, () => {
      const payload = {
        propOne: `one`,
        propTwo: {
          propOne: `two`
        }
      };

      const action = {
        type: types.LOAD_FILMS,
        payload
      };

      expect(reducer({}, action)).toEqual({
        data: payload
      });
      expect(reducer(initState, action)).toEqual({
        data: payload
      });
    });

    it(`Should return state on wrong action`, () => {
      const action = {};
      const store = {
        data: {
          allIDs: [],
          byIDs: {}
        }
      };

      expect(reducer(store, action)).toEqual(store);
      expect(reducer(initState, action)).toEqual(initState);
    });
  });
});

describe(`Reducers: Films selectors`, () => {
  it(`Selector getAllIDs`, () => {
    expect(selectors.getAllIDs(mocks.store)).toEqual([1, 3]);
  });

  it(`Selector getFilmsByIDs`, () => {
    expect(selectors.getFilmsByIDs(mocks.store)).toEqual(mocks.store.films.data.byIDs);
  });

  it(`Selector getAllFilmsGenres`, () => {
    expect(selectors.getAllFilmsGenres(mocks.store)).toEqual(
        [{
          id: 1,
          genre: `genreOne`,
        }, {
          id: 3,
          genre: `genreTwo`,
        }]
    );
  });

  it(`Selector getCurrentCardsInfo`, () => {
    expect(selectors.getCurrentCardsInfo.resultFunc([1, 3], mocks.store.films.data.byIDs)).toEqual(
        [{
          id: 1,
          name: `filmOne`,
          preview: {
            image: `img/filmOne.jpg`,
            videoSrc: `https://some-linkOne`
          }
        }, {
          id: 3,
          name: `filmTwo`,
          preview: {
            image: `img/filmTwo.jpg`,
            videoSrc: `https://some-linkTwo`
          }
        }]
    );

    expect(selectors.getCurrentCardsInfo.resultFunc([3], mocks.store.films.data.byIDs)).toEqual([{
      id: 3,
      name: `filmTwo`,
      preview: {
        image: `img/filmTwo.jpg`,
        videoSrc: `https://some-linkTwo`
      }
    }]);

    expect(selectors.getCurrentCardsInfo.resultFunc([], mocks.store.films.data.byIDs)).toEqual([]);
  });
});
