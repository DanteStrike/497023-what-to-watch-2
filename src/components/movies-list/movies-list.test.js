import React from "react";
import renderer from "react-test-renderer";
import MoviesList from "./movies-list.jsx";
import {films} from "../../mocks/films.js";

it(`Render correctly MoviesList component`, () => {
  const MoviePreviewComponent = renderer
    .create(
        <MoviesList
          films={films}
        />,
        {
          createNodeMock: (element) => {
            if (element.type === `video`) {
              return {
                src: null,
                isMuted: null,
                poster: null
              };
            }
            return null;
          }
        }
    ).toJSON();

  expect(MoviePreviewComponent).toMatchSnapshot();
});
