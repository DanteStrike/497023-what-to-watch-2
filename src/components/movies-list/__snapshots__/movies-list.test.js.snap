// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render correctly MoviesList component 1`] = `
<div
  className="catalog__movies-list"
>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "willChange": "transform",
      }
    }
  >
    <a
      href="/films/1"
      onClick={[Function]}
      style={
        Object {
          "color": "unset",
          "textDecoration": "none",
        }
      }
    >
      <div
        className="small-movie-card__image"
      >
        <video
          preload="none"
          src="src"
          style={
            Object {
              "display": "block",
              "height": "100%",
              "objectFit": "fill",
              "width": "100%",
            }
          }
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        name
      </h3>
    </a>
  </article>
</div>
`;
